### PR TITLE
Add panoramas to the new schema

### PR DESCRIPTION
**Pull request by WillNilges**
Successor to: https://github.com/WillNilges/meshdb/pull/216

### CODE REVIEW

**Comment by WillNilges** [comment]
should probably address comments from: https://github.com/WillNilges/meshdb/pull/216

**Comment by WillNilges** [comment]
Also we should look at Network Number and fall back to Install Number

File: pyproject.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ dependencies = [
     "drf-spectacular==0.27.*",
     "djangorestframework-dataclasses==1.3.*",
     "django-nonrelated-inlines==0.2.*",
+    "django-jsonform@git+https://github.com/willnilges/django-jsonform.git@51cbed42ccdaec81a35c97a919d054e2c9ca0207",
```
Comment (Andrew-Dickinson) [inline pyproject.toml:30]: Can you add a todo with a link to your PR?

File: src/meshapi/widgets.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class PanoramaViewer(JSONFormWidget):
+    pano_template_name = "widgets/panorama_viewer.html"
+
+    def __init__(self, schema):
+        super().__init__(schema)
+
+    def pano_get_context(self, name, value, attrs=None):
+        # FIXME: Need to parse the value because Django gives it as a string
```
Comment (Andrew-Dickinson) [inline src/meshapi/widgets.py:15]: Remove

File: src/meshapi/views/panoramas.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                    f"Warning: Could not add panorama to building (Install #{install_number}). Install does not exist."
+                )
+                warnings.append(install_number)
+                continue
+            for filename in filenames:
+                file_url = f"{host_url}{filename}"
+                panoramas.append(file_url)
+            if install.building.panoramas == panoramas:
+                continue
+            install.building.panoramas = panoramas
```
Comment (Andrew-Dickinson) [inline src/meshapi/views/panoramas.py:83]: Overwrites old installs

File: src/meshapi/views/panoramas.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    return master["commit"]["commit"]["tree"]["sha"]
+
+
+# Returns all the filenames, stripped of extensions and everything
+def list_files_in_git_directory(owner: str, repo: str, directory: str, tree):
+    url = f"https://api.github.com/repos/{owner}/{repo}/git/trees/{tree}?recursive=1"
+    token = os.environ.get("PANO_GITHUB_TOKEN")
+    response = requests.get(url, headers={"Authorization": f"Bearer {token}"})
+    if response.status_code != 200:
+        print(f"Error: Failed to fetch GitHub directory contents. Status code: {response.status_code}")
```
Comment (Andrew-Dickinson) [inline src/meshapi/views/panoramas.py:163]: Make me an exception

File: src/meshapi/widgets.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        value_as_array = json.loads(value)
+        return {
+            "widget": {
+                "name": name,
+                "value": value_as_array,
+            }
+        }
+
+    def render(self, name, value, attrs=None, renderer=None):
+        # Render the JSONFormWidget to allow editing of the panoramas
```
Comment (Andrew-Dickinson) [inline src/meshapi/widgets.py:26]: Disallow editing

File: src/meshapi/urls.py [hunk 6]
```diff
@@ -31,4 +31,5 @@
     path("mapdata/links/", views.MapDataLinkList.as_view(), name="meshapi-v1-map-data-links"),
     path("mapdata/sectors/", views.MapDataSectorList.as_view(), name="meshapi-v1-map-data-sectors"),
     path("geography/whole-mesh.kml", views.WholeMeshKML.as_view(), name="meshapi-v1-geography-whole-mesh-kml"),
+    path("update-panoramas/", views.update_panoramas_from_github, name="meshapi-v1-update-panoramas"),
```
Comment (Andrew-Dickinson) [inline src/meshapi/urls.py:34]: Run hourly with celery?

**Review by Andrew-Dickinson — APPROVED** [review]

File: src/meshapi/views/panoramas.py (hunk 4 above)
Comment (WillNilges) [reply src/meshapi/views/panoramas.py:163]: no :)

**Comment by WillNilges** [comment]
maybe I need to add tests for hourly celery and for the new panorama upsert logic

**Pull request merged**